### PR TITLE
Auth, caching, compression and multiprocessing fixes

### PR DESCRIPTION
**Pull request by allait**
### Disable caching for the metrics endpoint

For apps deployed behind a CDN we need to set cache control to prevent the CDN from caching the metrics response.

### Set prometheus_multiproc_dir before we import prometheus client

prometheus_client checks for the environment variable during import, so we need to set it before we load the library for the first time.

### Fix duplicate metrics in multiprocess mode

Prometheus client multiprocessing mode disables registries isolation, so all metrics are exported. Multiprocess collector is configured on one registry. If any metrics are explicitly connected to that registry they end up in the metrics output twice - once for the aggregate multiprocess value and once with the values from the individual worker process. To avoid this we don't connect any metrics to the registry and use the default one instead.

### Use request route expression instead of path

[Prometheus docs][cardinality-docs] recommend to keep the metric cardinality low and avoid using unique identifiers in the labels.

Depending on the application, using request path as a metric label could dramatically increase the amount of data stored.

To avoid this, we're using request route pattern instead of the actual path as the metric label (eg `/documents/<id>` instead of `/documents/123`).

This still allows to differentiate between different request types but since the number of endpoints in an app is generally limited it avoids creating thousands (or millions) of additional time series.

[cardinality-docs]: https://prometheus.io/docs/practices/instrumentation/#do-not-overuse-labels

### Move metrics definitions to a separate module

Since we don't need to set a registry for metrics in the multiprocess mode we can move them out of the class definition to a separate module.

### Add an authentication check to the metrics endpoint

Metrics endpoint requires an auth header with a bearer token to be provided when an app is run in a CloudFoundry environment.

Application ID is used as the bearer token value by the Prometheus scraper.

### Compress metrics endpoint response with gzip

Prometheus sets the 'Accept-Encoding: gzip' header when scraping application metrics, so we can gzip the metrics response to reduce the size.

### Label request without a matching rule with "No endpoint"

Since we're using request URL rule instead of the path to keep the number of labels manageable we need to handle 404 requests for URLs that don't exist in the app. `None` looks confusing in the metrics list, so we're replacing it with "No endpoint".

### Proxy imports for Prometheus client metric types

Allows users to create new metrics by importing the metric type from gds_metrics, without exposing the underlying prometheus client library.

### Add a child_exit function for gunicorn multiprocessing support

Prometheus client multiprocessing mode requires gunicorn to clean up temporary files to be cleaned up when workers exit, so we need to install the gunicorn child_exit hook.

Providing the function in a gds_metrics.gunicorn module should allow users to import it into their gunicorn config files without any additional changes.

### CODE REVIEW

**Comment by kentsanggds** [comment]
I've added some tests to make sure that the app is working as expected and a Makefile with other distributable files to make it easier to publish the package to PyPI.

File: scripts/run_tests.sh [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  fi
+}
+
+if [[ -z "$VIRTUAL_ENV" ]] && [[ -d venv ]]; then
+  source ./venv/bin/activate
+fi
+flake8 .
+display_result $? 1 "Code style check"
+
+py.test --cov=gds_metrics --cov-report=term-missing tests/ -v
```
Comment (leohemsted) [inline scripts/run_tests.sh]: i'd have a look into how we use tox in python client (which we mostly stole from looking at other libs like requests, flask, celery, etc)

recently changed in https://github.com/alphagov/notifications-python-client/pull/104

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+.PHONY: help
+help:
+	@cat $(MAKEFILE_LIST) | grep -E '^[a-zA-Z_-]+:.*?## .*$$' | sort | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
+
+.PHONY: venv
+venv: venv/bin/activate ## Create virtualenv if it does not exist
+
+venv/bin/activate:
+	test -d venv || virtualenv venv -p python3
+	./venv/bin/pip install pip-accel
```
Comment (leohemsted) [inline Makefile]: i really dislike these virtualenv commands - they don't work well with using mkvirtualenv. I'd just sort out the requirements ignoring what virtualenv does.

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+.PHONY: dependencies
+dependencies: venv ## Install build dependencies
+	mkdir -p ${PIP_ACCEL_CACHE}
+	PIP_ACCEL_CACHE=${PIP_ACCEL_CACHE} ./venv/bin/pip-accel install --upgrade -r requirements_for_test.txt
```
Comment (leohemsted) [inline Makefile]: pipaccel is kinda weird and i'm not sure it offers huge improvements for a small project like this - i'd probably skip it.

File: Makefile (hunk 2 above)
Comment (kentsanggds) [reply Makefile]: ok, sure

File: Makefile (hunk 3 above)
Comment (kentsanggds) [reply Makefile]: ok

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	./venv/bin/python setup.py bdist_wheel
+
+.PHONY: publish-to-pypi
+publish-to-pypi: build-wheel ## upload distributable wheel to pypi
+	./venv/bin/pip install --upgrade twine
+	@./venv/bin/twine upload dist/*.whl \
+		--repository-url=https://upload.pypi.org/legacy/ \
+		--username="${PYPI_USERNAME}" \
+		--password="${PYPI_PASSWORD}" \
+		--skip-existing # if you haven't run `make clean` there may be old wheels - don't try and re-upload them
```
[the remark this answers is not in the document]
Comment (kentsanggds) [reply Makefile]: at the moment it works, might need to revisit if there are updates to twine

File: scripts/run_tests.sh (hunk 1 above)
Comment (kentsanggds) [reply scripts/run_tests.sh]: ok, will try it out

**Pull request merged**